src/AppInstallerCLICore/ContextOrchestrator.cpp

```diff
@@ -0,0 +1,165 @@
+
+        // Check if item can be dequeued.
+        // Since only one item can be installed at a time currently the logic is very simple,
+        // and can just check if the first item is already running. This logic will need to become
+        // more complicated if multiple operation types (e.g. Download & Install) are added that can
+        // run simultaneously.
+        if (item->GetState() == OrchestratorQueueItemState::Running)
+        {
+            return {};
+        }
+
+        // Running state must be set inside the queueLock so that multiple threads don't try to run the same item.
+        item->SetState(OrchestratorQueueItemState::Running);
+        return item;
+    }
+
+    void ContextOrchestrator::RunContexts()
+    {
+        std::shared_ptr<OrchestratorQueueItem> item = GetNextItem();
+        while(item != nullptr)
+        {
+            HRESULT terminationHR = S_OK;
+            try
+            {
+                ::AppInstaller::CLI::RootCommand rootCommand;
+                std::unique_ptr<::AppInstaller::CLI::Command> command = std::make_unique<::AppInstaller::CLI::COMInstallCommand>(rootCommand.Name());
+                ::AppInstaller::Logging::Telemetry().LogCommand(command->FullName());
+                rootCommand.ValidateArguments(item->GetContext()->Args);
+
+                item->GetContext()->EnableCtrlHandler();
+
+                terminationHR = ::AppInstaller::CLI::Execute(*item->GetContext(), command);
+            }
+            WINGET_CATCH_STORE(terminationHR);
+
+            if (FAILED(terminationHR))
+            {
+                item->GetContext()->Terminate(terminationHR);
+            }
+
+            RemoveItemInState(item, OrchestratorQueueItemState::Running);
+
+            item = GetNextItem();
+        }
+    }
+
+    void ContextOrchestrator::RemoveItemInState(const std::shared_ptr<OrchestratorQueueItem>& item, OrchestratorQueueItemState state)
+    {
+        std::lock_guard<std::mutex> lock{ m_queueLock };
+
+        // Look for the item. It's ok if the item is not found since multiple listeners may try to remove the same item.
+        auto itr = std::find(m_queueItems.begin(), m_queueItems.end(), item);
+        if (itr != m_queueItems.end() && (*itr)->GetState() == state)
+        {
+            m_queueItems.erase(itr);
+
+            const auto& manifest = item->GetContext()->Get<Execution::Data::Manifest>();
+            m_installingWriteableSource->RemovePackageVersion(manifest, std::filesystem::path{ manifest.Id + '.' + manifest.Version });
+
+            ::SetEvent(item->GetCompletedEvent());
+        }
+    }
+    void ContextOrchestrator::CancelItem(const std::shared_ptr<OrchestratorQueueItem>& item)
+    {
+        // Always cancel the item, even if it isn't running yet, to get the terminationHR set correctly.
+        item->GetContext()->Cancel(false, true);
+
+        RemoveItemInState(item, OrchestratorQueueItemState::Queued);

```

Comment:
Yes, it can't be removed early. The problem is that Cancel doesn't actually work reliably (i.e. if the install is running)or synchronously, so this would mean that once Cancel was called we wouldn't be tracking apps whose installer was running. But the queue is supposed to prevent multiple installers from running at a time.